### PR TITLE
Allow users to pass task payload via deep storage instead of environment variable

### PR DESCRIPTION
**Pull request by georgew5656**
This change is meant to fix a issue where passing too large of a task payload to the mm-less task runner will cause the peon to fail to startup because the payload is passed (compressed) as a environment variable (TASK_JSON). In linux systems the limit for a environment variable is commonly 128KB, for windows systems less than this. Setting a env variable longer than this results in a bunch of "Argument list too long" errors.

### Description
**(1) Problem**
The goal of this patch is to prevent larger tasks from failing with mm-less ingestion due to the TASK_JSON being too large as described above.

**(2) Solution**
Part 1. Optionally stop setting TASK_JSON

To address the immediate problem (setting environment variables that are too large), I added a additional config for the KubernetesTaskRunner (druid.indexer.runner.taskPayloadAsEnvVariable) that defaults to true but can be optionally set to false. Setting this config to false will cause the K8s adapters to not set the task payload as the TASK_JSON env variable. This prevents the Jobs from failing to come up.

Part 2. We still need to pass the task.json payload to the peons somehow. I explored three options for this, and ended up going with the below solution.

Push the task payload into task logs deep storage and have the peon read the payload.
I ended up going with this option because it was the most simple to implement and the most future-proof (no worry about task payloads getting larger than 1MB). The task logs killer will automatically delete the task.json in deep storage alongside he task logs whenever it is run.

Changes Made
- Introduced a new interface (TaskPayloadManager) that exposes two methods (push task payload, pull task payload) and is implemented by TaskLogs. (I only implemented it for S3TaskLogs, but it should be easy to implement for other deep storage systems).
- Introduce a new config in TaskConfig (druid.indexer.task.enableTaskPayloadManagerPerTask). When set on the overlord, mm-less ingestion will push the task.json payload to deep storage before launching a k8s job. When set on the peon, the CliPeon task injector will read the task.json from deep storage instead of assuming it is available on the file system. This is currently only useable by mm-less ingestion but it technically could be used in any task running scenario.
- The K8s adapters in mm-less ingestion will check deep storage if the config is set when converting K8s jobs to tasks in toTask

**(3) Alternative solutions to passing the task.json payload**
1. Using k8s configmaps to store the task payload and then mounting them onto the created peon pods. 
I decided not to go with this option because configmaps still have a 1MB size limit and I was concerned with the KubernetesTaskRunner having to manage a bunch of configmaps in addition to jobs. Having this many configmaps also pollutes K8s metadata, making it hard to see anything else going on when you're looking at configmaps.

2. Updating CliPeon to use the getTaskPayload endpoint on the overlord to pull the task.json payload on startup. This didn't work because we currently have a guice injector in the peon that requires the task.json be available at injection time. In order to pull the task.json from the overlord, we need to use the ServiceLocator class which is only available once the peon lifecycle has already started (after injection). Changing this would have required many changes to the code so I didn't want to do it. Additionally, I would have had to deprecate the toTask interface on the overlords since there would be no way for the overlord to turn a K8s Job into a task definition.

#### Release note
- Support compressed task payloads larger than 128KB in mm-less ingestion.

##### Key changed/added classes in this PR
 * `CliPeon`
 * `KubernetesPeonLifecycle `
 * `PodTemplateTaskAdapter`
 * `K8sTaskAdapter`
 * `S3TaskLogs` `TaskLogs`

I can add some more documentation to this PR later but I wanted to get some feedback on this approach before doing so.

This PR has:

- [X] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [X] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/TaskAdapter.java
Comment (YongGang) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/TaskAdapter.java:35]: Seems this method has same implementation in the subclasses, maybe move here as a default method.

File: services/src/main/java/org/apache/druid/cli/CliPeon.java
Comment (YongGang) [inline services/src/main/java/org/apache/druid/cli/CliPeon.java:293]: What's the reason to use a Smile mapper? Seems same as Json mapper.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java
Comment (YongGang) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java:160]: The nested try-catch seems can be simplified.

**Comment by cryptoe** [comment]
@georgew5656
The current approach also SGTM but I would not have if behind a feature flag. If you want to have a feature flag, I would optionally turn it on by default .
This also means we implement that TaskPayloadManager for other deep storage mainly 'GCS,AZURE,HDFS`. If the implementation is not found, we should through a nice error message instructing the user to set `druid.indexer.runner.taskPayloadAsEnvVariable` to true or if we are not using a flag, instructing the user to implement the methods to implement 'TaskPayloadManager` for there custom deep storage impl.

**Comment by churromorales** [comment]
@cryptoe I think the way it is implemented now behind the feature flag is better.  For most usecases it is much better to just pass the task.json directly.  It is much faster, also for our customers we use deep-storage which is slow and less featured than s3, thus this would be a feature we would only turn on if necessary.  Additionally, our deep storage provider wont allow us to do batch deletes so the cleaner approach to remove the task files is not great.  I know its a special case for us, but it is always better to pass something directly if possible than use indirection...disregarding our usecase.

LGTM overall, but one feature request I would like to request is that at the end of the task in AbstractTask to delete the task.json file, not leave it up to the cleaner.  Quite a few folks that have their own k8s which they launch in their datacenters.   While this works for cloud providers, it might not work for everyone else.  This can be done in another PR or I can one up after this PR is merged if needed.

File: services/src/main/java/org/apache/druid/cli/CliPeon.java
Comment (georgew5656) [reply services/src/main/java/org/apache/druid/cli/CliPeon.java:293]: we use the smile mapper to serialize the payload so i figured we should use the same one to deserialize

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java
Comment (georgew5656) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java:160]: how do you reccommend simplifying this? the createTempFile can throw a IOException

**Comment by georgew5656** [comment]
> @georgew5656 The current approach also SGTM but I would not have it behind a feature flag. If you want to have a feature flag, I would optionally turn it on by default . This also means we implement that TaskPayloadManager for other deep storage mainly 'GCS,AZURE,HDFS`. If the implementation is not found, we should through a nice error message instructing the user to set `druid.indexer.runner.taskPayloadAsEnvVariable` to true or if we are not using a flag, instructing the user to implement the methods to implement 'TaskPayloadManager` for there custom deep storage impl.

i didn't really want to break anyone who didn't want to use deep storage for task payloads for whatever reason. i think maybe later on if this gets used in production a bit more and the performance is okay maybe we could consider flipping it on as a default

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/TaskAdapter.java
Comment (georgew5656) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/TaskAdapter.java:35]: i didn't want to do this because the implementation depends on logic in fromTask, since both adapters happen to add the same annotation for taskId the implementation for getTaskId is the same but this isn't necessarily the case

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java
Comment (YongGang) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java:160]: Don't think we need to handle the IOException explicitly as it is wrapped to RuntimeException anyway.
So the code can be like:
```
      Path file = null;
      try {
        file = Files.createTempFile(taskId.getOriginalTaskId(), "task.json");
        FileUtils.writeStringToFile(file.toFile(), mapper.writeValueAsString(task), Charset.defaultCharset());
        taskLogs.pushTaskPayload(task.getId(), file.toFile());
      }
      catch (Exception e) {
        log.error("Failed to write task payload for task: %s", taskId.getOriginalTaskId());
        throw new RuntimeException(e);
      }
      finally {
        if (file ! = null) {
          Files.deleteIfExists(file);
        }
      }
    }
```

**Comment by cryptoe** [comment]
Thanks @churromorales and @georgew5656 for the responses. I was more worried about adding another config which needs to be set for the end users. 
The current opt out by default lgtm.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunnerConfig.java
Comment (YongGang) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunnerConfig.java:422]: I don't see where this method is called. 
And also could we consolidate the two new config items to one? The current two configs are excluding each other thus can be misconfigured. Maybe like this:
| Config Name | Type | Description | Default | Required |
|-------------|------|-------------|---------|----------|
| `druid.indexer.task.taskPayloadHandlingMethod` | `String` | Determines how task payloads should be passed to K8s jobs. Use `ENV_VARIABLE` if you want task payloads to be passed as a compressed environment variable (suitable for payloads <= 1 MB). Use `PAYLOAD_MANAGER` if you plan on using task payloads > 1 MB. | `ENV_VARIABLE` | No |

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java
Comment (georgew5656) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java:160]: the issue is the deleteIfExists call throws a IOException so I didn't want to have the whole function do that

**Comment by abhishekagarwal87** [comment]
I have concerns about the config change as well. In a single PR adding one config doesn't seem much, but after a year worth of work, you suddenly realize that the feature has become very complex to tune and use. Is there a way to check if the deep storage supports storing payload and if not, then using the environment variable? Because then we can change the default easily.

**Comment by churromorales** [comment]
Why not just check the size of the task.json, if its larger than then MAX_SIZE do it in deep storage, if it is smaller then just use the env? I agree 100% with Druid in general having too many configuration options, makes it hard to remember everything you need include.  When creating this feature, the whole goal I had in mind was to have this work with as few configuration options as possible.

**Comment by georgew5656** [comment]
> Why not just check the size of the task.json, if its larger than then MAX_SIZE do it in deep storage, if it is smaller then just use the env? I agree 100% with Druid in general having too many configuration options, makes it hard to remember everything you need include. When creating this feature, the whole goal I had in mind was to have this work with as few configuration options as possible.

i think MAX_SIZE depends on the underlying OS but I don't mind using the 100KB default for linux and doing it that way

**Comment by georgew5656** [comment]
@abhishekagarwal87 @churromorales I just put up a new version of the PR that removes the configs and uses deep storage to pass the task payload if the task size is too large. This is more of a "opinionated" choice but in this case I think it makes sense, lmk what you guys think

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3TaskLogs.java
Comment (abhishekagarwal87) [inline extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3TaskLogs.java:90]:
```suggestion
    log.info("Pushing task payload [%s] to location [%s]", taskPayloadFile, taskKey);
```

File: processing/src/main/java/org/apache/druid/tasklogs/TaskPayloadManager.java
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/tasklogs/TaskPayloadManager.java:32]: The javadocs can be more descriptive. Like how it is used, where it is used etc. "manage task json payloads" is very broad description. It's only used in k8s task management.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java
Comment (georgew5656) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java:157]: the task reads the file to disk on startup so i wouldn't be worried about the log cleanup job to clean it up that soon. we are relying on the log cleanup job to cleanup deep storage

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java
Comment (georgew5656) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java:122]: none really

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java
Comment (georgew5656) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java:153]: the druid exception classes all look like query exceptions to me, are u suggesting adding a new internal exception or something?

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java
Comment (georgew5656) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java:153]: i would still have to add a IOException to method signature since the delete would happen in the finally block, but maybe that's okay?

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java
Comment (georgew5656) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java:486]: the compiler is forcing the call, but I guess we can make this method throw IOException since all the other methods in the interface do

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java
Comment (abhishekagarwal87) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesPeonLifecycle.java:153]: Yeah. I think so. Unless you don't want to handle it up the stack.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java
Comment (abhishekagarwal87) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java:153]: They are not just query exceptions. They are used for generating exceptions that will be returned to the end user. So it could either be a query user, an ingestion user, or a cluster admin.

**Comment by abhishekagarwal87** [comment]
thank you for addressing comments @georgew5656. Looks good to me except for the exception handling in some places.

File: processing/src/main/java/org/apache/druid/error/InternalError.java
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/error/InternalError.java:60]: it should be runtime_failure category.

File: processing/src/main/java/org/apache/druid/error/InternalError.java
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/error/InternalError.java:22]: how about InternalServerError?

File: processing/src/main/java/org/apache/druid/error/InternalError.java
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/error/InternalError.java:50]: Why canceled? Instead, you can force the caller to pass on an error code.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java
Comment (abhishekagarwal87) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java:154]: is there an action you can associate with this error message? Like should they verify that overlord is successfully uploading task jsons to deep storage.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java
Comment (abhishekagarwal87) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java:165]: can this be replaced with DruidException.defensive()?

File: processing/src/main/java/org/apache/druid/error/InternalError.java
Comment (georgew5656) [reply processing/src/main/java/org/apache/druid/error/InternalError.java:50]: i just figured this was the default 500 error code

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java
Comment (georgew5656) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java]: updated the message

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java
Comment (abhishekagarwal87) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java:155]:
```suggestion
          "Could not load task payload from deep storage for job [%s]. Check the overlord logs for any errors in uploading task payload to deep storage.",
```

**Review by abhishekagarwal87 — COMMENTED** [review]
Almost there. Can you look into the test failures?

**Comment by georgew5656** [comment]
> Almost there. Can you look into the test failures?

i think the only failing tests are for coverage now so we should be good

**Review by abhishekagarwal87 — APPROVED** [review]

**Comment by abhishekagarwal87** [comment]
@georgew5656 - I will let you merge with a nicer commit summary that what github is generating.

**Pull request merged**
